File: src/maze.py
"""
迷宫模块 - 用于生成和表示迷宫
"""
import random
import numpy as np
from typing import Tuple, List


class Maze:
    """迷宫类，用于表示和操作迷宫"""
    
    def __init__(self, width: int, height: int):
        """
        初始化迷宫
        
        Args:
            width: 迷宫宽度
            height: 迷宫高度
        """
        self.width = width
        self.height = height
        self.grid = np.ones((height, width), dtype=int)  # 1表示墙，0表示路
        self.start = (1, 1)  # 起点
        self.end = (height - 2, width - 2)  # 终点
        
    def generate_maze(self, complexity: float = 0.75, density: float = 0.75):
        """
        使用随机算法生成迷宫
        
        Args:
            complexity: 迷宫复杂度 (0-1)
            density: 迷宫密度 (0-1)
        """
        # 创建边界
        self.grid = np.zeros((self.height, self.width), dtype=int)
        self.grid[0, :] = self.grid[-1, :] = 1
        self.grid[:, 0] = self.grid[:, -1] = 1
        
        # 生成内部墙壁
        complexity = int(complexity * (5 * (self.height + self.width)))
        density = int(density * ((self.height // 2) * (self.width // 2)))
        
        for _ in range(density):
            x = random.randint(0, self.width // 2) * 2
            y = random.randint(0, self.height // 2) * 2
            self.grid[y, x] = 1
            
            for _ in range(complexity):
                neighbours = []
                if x > 1:
                    neighbours.append((y, x - 2))
                if x < self.width - 2:
                    neighbours.append((y, x + 2))
                if y > 1:
                    neighbours.append((y - 2, x))
                if y < self.height - 2:
                    neighbours.append((y + 2, x))
                    
                if neighbours:
                    y_, x_ = neighbours[random.randint(0, len(neighbours) - 1)]
                    if self.grid[y_, x_] == 0:
                        self.grid[y_, x_] = 1
                        self.grid[y_ + (y - y_) // 2, x_ + (x - x_) // 2] = 1
                        x, y = x_, y_
        
        # 确保起点和终点是通路
        self.grid[self.start[0], self.start[1]] = 0
        self.grid[self.end[0], self.end[1]] = 0
        
    def is_valid_position(self, position: Tuple[int, int]) -> bool:
        """
        检查位置是否有效（在边界内且不是墙）
        
        Args:
            position: (y, x) 坐标
            
        Returns:
            是否有效
        """
        y, x = position
        if 0 <= y < self.height and 0 <= x < self.width:
            return self.grid[y, x] == 0
        return False
    
    def get_valid_neighbors(self, position: Tuple[int, int]) -> List[Tuple[int, int]]:
        """
        获取位置的所有有效邻居
        
        Args:
            position: (y, x) 坐标
            
        Returns:
            有效邻居列表
        """
        y, x = position
        neighbors = [
            (y - 1, x),  # 上
            (y + 1, x),  # 下
            (y, x - 1),  # 左
            (y, x + 1),  # 右
        ]
        return [n for n in neighbors if self.is_valid_position(n)]
    
    def calculate_distance(self, pos1: Tuple[int, int], pos2: Tuple[int, int]) -> float:
        """
        计算两点之间的曼哈顿距离
        
        Args:
            pos1: 第一个位置
            pos2: 第二个位置
            
        Returns:
            曼哈顿距离
        """
        return abs(pos1[0] - pos2[0]) + abs(pos1[1] - pos2[1])
    
    def __str__(self) -> str:
        """返回迷宫的字符串表示"""
        result = []
        for y in range(self.height):
            row = []
            for x in range(self.width):
                if (y, x) == self.start:
                    row.append('S')
                elif (y, x) == self.end:
                    row.append('E')
                elif self.grid[y, x] == 1:
                    row.append('█')
                else:
                    row.append(' ')
            result.append(''.join(row))
        return '\n'.join(result)

File: web/src/maze/Maze.ts
// 迷宫类
import type { MazeData, Position } from '../types';

export class Maze {
  grid: number[][];
  start: Position;
  end: Position;
  width: number;
  height: number;

  constructor(
    width: number,
    height: number
  ) {
    this.width = width;
    this.height = height;
    this.grid = Array(height).fill(0).map(() => Array(width).fill(1));
    this.start = [1, 1];
    this.end = [height - 2, width - 2];
  }

  static fromJSON(data: MazeData): Maze {
    const maze = new Maze(data.width, data.height);
    maze.grid = data.grid;
    maze.start = data.start;
    maze.end = data.end;
    return maze;
  }

  generate(complexity: number = 0.85, density: number = 0.85, seed?: number): void {
    // 初始化
    this.grid = Array(this.height).fill(0).map(() => Array(this.width).fill(0));
    
    // 边界
    for (let i = 0; i < this.width; i++) {
      this.grid[0][i] = 1;
      this.grid[this.height - 1][i] = 1;
    }
    for (let i = 0; i < this.height; i++) {
      this.grid[i][0] = 1;
      this.grid[i][this.width - 1] = 1;
    }

    // 随机生成墙壁
    const rng = seed !== undefined ? this.seededRandom(seed) : Math.random;
    const complexityVal = Math.floor(complexity * 5 * (this.height + this.width));
    const densityVal = Math.floor(density * (this.height / 2) * (this.width / 2));

    for (let i = 0; i < densityVal; i++) {
      let x = Math.floor(rng() * (this.width / 2)) * 2;
      let y = Math.floor(rng() * (this.height / 2)) * 2;
      this.grid[y][x] = 1;

      for (let j = 0; j < complexityVal; j++) {
        const neighbors: Position[] = [];
        if (x > 1) neighbors.push([y, x - 2]);
        if (x < this.width - 2) neighbors.push([y, x + 2]);
        if (y > 1) neighbors.push([y - 2, x]);
        if (y < this.height - 2) neighbors.push([y + 2, x]);

        if (neighbors.length > 0) {
          const [ny, nx] = neighbors[Math.floor(rng() * neighbors.length)];
          if (this.grid[ny][nx] === 0) {
            this.grid[ny][nx] = 1;
            this.grid[y + Math.floor((ny - y) / 2)][x + Math.floor((nx - x) / 2)] = 1;
            x = nx;
            y = ny;
          }
        }
      }
    }

    // 确保起点和终点可通行
    this.grid[this.start[0]][this.start[1]] = 0;
    this.grid[this.end[0]][this.end[1]] = 0;
  }

  private seededRandom(seed: number): () => number {
    return () => {
      seed = (seed * 9301 + 49297) % 233280;
      return seed / 233280;
    };
  }

  isValid(pos: Position): boolean {
    const [y, x] = pos;
    return y >= 0 && y < this.height && x >= 0 && x < this.width && this.grid[y][x] === 0;
  }

  getNeighbors(pos: Position): Position[] {
    const [y, x] = pos;
    const neighbors: Position[] = [
      [y - 1, x],
      [y + 1, x],
      [y, x - 1],
      [y, x + 1]
    ];
    return neighbors.filter(n => this.isValid(n));
  }

  manhattanDistance(pos1: Position, pos2: Position): number {
    return Math.abs(pos1[0] - pos2[0]) + Math.abs(pos1[1] - pos2[1]);
  }

  toJSON(): MazeData {
    return {
      width: this.width,
      height: this.height,
      grid: this.grid,
      start: this.start,
      end: this.end
    };
  }
}

